File: utils/udp_object_detection.py
#!/usr/bin/env python3
#
# SPDX-License-Identifier: BSD-2-Clause
#
# Copyright (C) 2025, Raspberry Pi Ltd.
#
# udp_object_detection.py - This utility file implements a UDP receiver
# designed to process object detection information. It's specifically
# tailored to receive data sent by `object_detect_udp_stage.cpp`,
# but can also be used with generic UDP senders like `nc -ulp 12347`
# for debugging raw data.

import socket
import struct
import sys
from dataclasses import dataclass

# --- Data Structures ---


@dataclass
class ParsedDetection:
    """A data class to hold the parsed detection data received over UDP."""
    x: int
    y: int
    width: int
    height: int
    name: str
    confidence: float

# --- UDP_AI_Receiver Class Definition ---


class UDP_AI_Receiver:
    """
    The UDP_AI_Receiver class handles the establishment of a UDP socket,
    receiving incoming data, and parsing it into the ParsedDetection format.
    """
    MAX_BUFFER_SIZE = 1024

    def __init__(self, port: int):
        """
        Constructor: Initializes the UDP receiver.
        :param port: The UDP port number to listen on.
        """
        try:
            # Create a UDP socket.
            # AF_INET specifies the IPv4 address family.
            # SOCK_DGRAM specifies a UDP (datagram) socket.
            self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            # Bind the created socket to the specified IP address and port.
            self.sock.bind(('', port))
        except socket.error as e:
            print(f"Failed to create or bind socket: {e}", file=sys.stderr)
            sys.exit(1)

        print(f"Listening for UDP detection packets on port {port}...")

    def __del__(self):
        """Destructor: Cleans up resources by closing the socket."""
        self.sock.close()

    def receive_detection(self) -> ParsedDetection | None:
        """
        Method to receive and parse a single detection packet.
        :return: A ParsedDetection object if a packet was successfully received and parsed,
                 None otherwise.
        """
        try:
            buffer, addr = self.sock.recvfrom(self.MAX_BUFFER_SIZE)
            return self._parse_detection(buffer)
        except socket.error as e:
            print(f"Failed to receive UDP message: {e}", file=sys.stderr)
            return None

    @staticmethod
    def _parse_detection(buffer: bytes) -> ParsedDetection | None:
        """
        Helper method to parse the raw byte buffer into a ParsedDetection object.
        :param buffer: A byte string containing the received data.
        :return: A ParsedDetection object if parsing was successful, None otherwise.
        """
        START_DELIMITER_LE = 0xDDCCBBAA

        # Define the minimum expected size of a valid packet, excluding the name string length.
        # Delimiter (4) + x (4) + y (4) + width (4) + height (4) + name_length (1) + confidence (4)
        MIN_PACKET_SIZE = 4 + 4 + 4 + 4 + 4 + 1 + 4

        if len(buffer) < MIN_PACKET_SIZE:
            print("Packet too small to be a valid detection.", file=sys.stderr)
            return None

        # Check for the presence of the start delimiter.
        delimiter = struct.unpack('<I', buffer[0:4])[0]
        if delimiter != START_DELIMITER_LE:
            print("Invalid packet: delimiter not found.", file=sys.stderr)
            return None

        offset = 4  # Start after the delimiter

        try:
            # Unpack the integer coordinates and dimensions.
            x, y, width, height = struct.unpack('<iiii', buffer[offset:offset + 16])
            offset += 16

            # Unpack the length of the object's name.
            name_length = struct.unpack('<B', buffer[offset:offset + 1])[0]
            offset += 1

            # Check for a malformed packet where the reported name length
            # would exceed the received buffer size.
            if offset + name_length + 4 > len(buffer):
                print("Invalid packet: name length exceeds buffer size.",
                      file=sys.stderr)
                return None

            # Unpack the name string.
            name = buffer[offset:offset + name_length].decode('utf-8')
            offset += name_length

            # Unpack the confidence score.
            confidence = struct.unpack('<f', buffer[offset:offset + 4])[0]

            return ParsedDetection(x, y, width, height, name, confidence)

        except struct.error as e:
            print(f"Error unpacking packet data: {e}", file=sys.stderr)
            return None

# --- Example Usage of the Receiver Class ---


if __name__ == "__main__":
    port = 12347  # The UDP port number this receiver will listen on.
    # Ensure this matches the port your sender is transmitting to.
    receiver = UDP_AI_Receiver(port)

    # Enter an infinite loop to continuously receive and process detection packets.
    while True:
        detection = receiver.receive_detection()
        if detection:
            # If a packet was successfully received and parsed, print its contents.
            print("Received Detection:")
            print(
                f"  Box: ({detection.x}, {detection.y}, {detection.width}, {detection.height})")
            print(f"  Name: {detection.name}")
            print(f"  Confidence: {detection.confidence}")

File: utils/udp_object_detection.cpp
/* SPDX-License-Identifier: BSD-2-Clause */
/*
 * Copyright (C) 2020, Raspberry Pi (Trading) Ltd.
 *
 * udp_object_detection.cpp - This utility file implements a UDP receiver
 * designed to process object detection information. It's specifically
 * tailored to receive data sent by `object_detect_draw_cv_stage.cpp`,
 * but can also be used with generic UDP senders like `nc -ulp 12347`
 * for debugging raw data.
 *
 * To compile this file:
 * g++ udp_object_detection.cpp -o udp_object_detection
 */

#include <iostream>  // For input/output operations (e.g., std::cout, std::cerr)
#include <string>    // For std::string manipulation (e.g., object name)
#include <vector>    // Not directly used in this version but often useful for collections
#include <sstream>   // Not directly used in this version but useful for string parsing/formatting
#include <cstring>   // For memcpy and memset, used for memory operations

// Required for networking functionalities
#include <sys/socket.h>  // For socket creation and manipulation
#include <netinet/in.h>  // For Internet address structures (e.g., sockaddr_in)
#include <unistd.h>      // For close() function
#include <arpa/inet.h>   // For INADDR_ANY and other network address conversions

// --- Data Structures ---

// A struct to hold the parsed detection data received over UDP.
// This structure defines the format of the object detection information.
struct ParsedDetection {
    int x, y;          // Top-left corner coordinates of the detected object's bounding box
    int width, height; // Dimensions of the detected object's bounding box
    std::string name;  // The name or label of the detected object (e.g., "person", "car")
    float confidence;  // The confidence score of the detection (0.0 to 1.0)
};

// --- UDP_AI_Receiver Class Definition ---

// The UDP_AI_Receiver class handles the establishment of a UDP socket,
// receiving incoming data, and parsing it into the ParsedDetection format.
class UDP_AI_Receiver {
public:
    // Constructor: Initializes the UDP receiver.
    // @param port The UDP port number to listen on.
    UDP_AI_Receiver(int port) {
        // Create a UDP socket.
        // AF_INET specifies the IPv4 address family.
        // SOCK_DGRAM specifies a UDP (datagram) socket.
        // 0 specifies the default protocol (UDP in this case).
        if ((sockfd_ = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
            perror("Failed to create socket"); // Print error message if socket creation fails
            exit(EXIT_FAILURE);                 // Terminate the program
        }

        // Initialize the server address structure with zeros.
        memset(&servaddr_, 0, sizeof(servaddr_));
        // Set the address family to IPv4.
        servaddr_.sin_family = AF_INET;
        // Bind to any available network interface. INADDR_ANY allows receiving
        // packets sent to any of the host's IP addresses.
        servaddr_.sin_addr.s_addr = INADDR_ANY;
        // Convert the port number to network byte order and assign it.
        servaddr_.sin_port = htons(port);

        // Bind the created socket to the specified IP address and port.
        // This makes the socket listen for incoming UDP packets on that port.
        if (bind(sockfd_, (const struct sockaddr *)&servaddr_, sizeof(servaddr_)) < 0) {
            perror("Failed to bind socket"); // Print error message if binding fails
            exit(EXIT_FAILURE);                // Terminate the program
        }
    }

    // Destructor: Cleans up resources when the UDP_AI_Receiver object is destroyed.
    ~UDP_AI_Receiver() {
        // Close the socket to release system resources.
        close(sockfd_);
    }

    // Method to receive and parse a single detection packet.
    // @param detection_data A reference to a ParsedDetection struct where the received
    //                       and parsed data will be stored.
    // @return true if a detection packet was successfully received and parsed, false otherwise.
    bool receiveDetection(ParsedDetection& detection_data) {
        char buffer[MAX_BUFFER_SIZE]; // Buffer to hold incoming UDP data
        socklen_t len = sizeof(clientaddr_); // Size of the client address structure

        // Receive data from the socket.
        // The received data is placed into 'buffer'.
        // 'n' will store the number of bytes received.
        ssize_t n = recvfrom(sockfd_, (char*)buffer, MAX_BUFFER_SIZE, 0,
                             (struct sockaddr *)&clientaddr_, &len);

        if (n < 0) {
            perror("Failed to receive UDP message"); // Print error if reception fails
            return false;
        }

        // Attempt to parse the received data.
        return parseDetection(buffer, n, detection_data);
    }

private:
    // Helper method to parse the raw byte buffer into a ParsedDetection struct.
    // @param buffer A pointer to the raw byte buffer containing the received data.
    // @param size The number of bytes in the buffer.
    // @param detection_data A reference to a ParsedDetection struct to populate.
    // @return true if parsing was successful, false otherwise (e.g., malformed packet).
    bool parseDetection(const char* buffer, size_t size, ParsedDetection& detection_data) {
        // Define a start delimiter used to validate the beginning of a valid packet.
        // This helps in discarding malformed or unrelated UDP traffic.
        const uint32_t START_DELIMITER = 0xDDCCBBAA;
        // Define the minimum expected size of a valid packet, excluding the name string length.
        const size_t MIN_PACKET_SIZE = sizeof(START_DELIMITER) + sizeof(int32_t) * 4 + sizeof(uint8_t) + sizeof(float);

        // Check if the received packet is too small to contain basic detection data.
        if (size < MIN_PACKET_SIZE) {
            std::cerr << "Packet too small to be a valid detection." << std::endl;
            return false;
        }

        // Check for the presence of the start delimiter at the beginning of the packet.
        // This ensures the packet adheres to the expected format.
        if (*(uint32_t*)buffer != START_DELIMITER) {
            std::cerr << "Invalid packet: delimiter not found." << std::endl;
            return false;
        }

        size_t offset = sizeof(START_DELIMITER); // Initialize offset after the delimiter

        // Read x, y, width, and height (each as a 32-bit integer).
        // Data is directly cast and read from the buffer at the current offset.
        detection_data.x = *(int32_t*)(buffer + offset);
        offset += sizeof(int32_t);
        detection_data.y = *(int32_t*)(buffer + offset);
        offset += sizeof(int32_t);
        detection_data.width = *(int32_t*)(buffer + offset);
        offset += sizeof(int32_t);
        detection_data.height = *(int32_t*)(buffer + offset);
        offset += sizeof(int32_t);

        // Read the length of the object's name (as an 8-bit unsigned integer).
        uint8_t name_length = *(uint8_t*)(buffer + offset);
        offset += sizeof(uint8_t);

        // Check if the reported name length would exceed the received buffer size,
        // which indicates a corrupted or malicious packet.
        if (offset + name_length + sizeof(float) > size) {
            std::cerr << "Invalid packet: name length exceeds buffer size." << std::endl;
            return false;
        }
        // Extract the name string using the `assign` method of `std::string`,
        // which copies `name_length` characters starting from `buffer + offset`.
        detection_data.name.assign(buffer + offset, name_length);
        offset += name_length;

        // Read the confidence score (as a 32-bit float).
        detection_data.confidence = *(float*)(buffer + offset);

        return true; // Parsing successful
    }

private:
    int sockfd_;                        // Socket file descriptor
    struct sockaddr_in servaddr_;       // Server address structure
    struct sockaddr_in clientaddr_;     // Client address structure (for `recvfrom` to store sender's info)
    static const int MAX_BUFFER_SIZE = 1024; // Maximum size for the UDP receive buffer
};

// --- Example Usage of the Receiver Class ---

int main() {
    int port = 12347; // The UDP port number this receiver will listen on.
                      // Ensure this matches the port your sender is transmitting to.
    UDP_AI_Receiver receiver(port); // Create an instance of the UDP_AI_Receiver

    std::cout << "Listening for UDP detection packets on port " << port << "..." << std::endl;

    // Enter an infinite loop to continuously receive and process detection packets.
    while (true) {
        ParsedDetection detection; // Declare a ParsedDetection struct to hold the received data
        if (receiver.receiveDetection(detection)) { // Attempt to receive and parse a packet
            // If a packet was successfully received and parsed, print its contents.
            std::cout << "Received Detection:" << std::endl;
            std::cout << "  Box: (" << detection.x << ", " << detection.y << ", " << detection.width << ", " << detection.height << ")" << std::endl;
            std::cout << "  Name: " << detection.name << std::endl;
            std::cout << "  Confidence: " << detection.confidence << std::endl;
        }
        // In a real-world application, you might add a small delay here
        // (e.g., `usleep(10000);`) to prevent the loop from consuming 100% CPU
        // if packets are not arriving continuously.
    }

    return 0; // Program exits (though in the infinite loop, this line is unreachable)
}